db/db_impl.cc

```diff
@@ -630,7 +635,97 @@ void DBImpl::StartTimedTasks() {
             stats_dump_period_sec * 1000000));
       }
     }
+    stats_persist_period_sec = mutable_db_options_.stats_persist_period_sec;
+    if (stats_persist_period_sec > 0) {
+      if (!thread_persist_stats_) {
+        thread_persist_stats_.reset(new rocksdb::RepeatableThread(
+            [this]() { DBImpl::PersistStats(); }, "pst_st", env_,
+            stats_persist_period_sec * 1000000));
+      }
+    }
+  }
+}
+
+void DBImpl::PersistStats() {
+  TEST_SYNC_POINT("DBImpl::PersistStats:1");
+#ifndef ROCKSDB_LITE
+  const DBPropertyInfo* cf_property_info =
+      GetPropertyInfo(DB::Properties::kCFStats);
+  assert(cf_property_info != nullptr);
+  const DBPropertyInfo* db_property_info =
+      GetPropertyInfo(DB::Properties::kDBStats);
+  assert(db_property_info != nullptr);
+  std::map<std::string, std::string> stats_map;
+  if (shutdown_initiated_) {
+    return;
+  }
+  {
+    InstrumentedMutexLock l(&mutex_);
+    default_cf_internal_stats_->GetMapProperty(
+        *db_property_info, DB::Properties::kDBStats, &stats_map);
+    std::map<std::string, std::string> cf_stats_map;
+    for (auto cfd : *versions_->GetColumnFamilySet()) {
+      if (cfd->initialized()) {
+        cfd->internal_stats()->GetMapProperty(
+            *cf_property_info, DB::Properties::kCFStatsNoFileHistogram,
+            &cf_stats_map);
+        cfd->internal_stats()->GetMapProperty(
+            *cf_property_info, DB::Properties::kCFFileHistogram, &cf_stats_map);
+        // add column family name to the key string
+        for (const auto& cf_stats : cf_stats_map) {
+          std::string key_with_cf_name = cfd->GetName() + "." + cf_stats.first;
+          stats_map[key_with_cf_name] = cf_stats.second;
```

Comment:
Am I right that the stats are accumulative counters? I believe we want the counters for the current window.

Reply:
`stats_map` is a single level std::map that stores the stats kv pairs. In this loop we augment the stats name to include the column family name. Once `stats_map` is populated, it will be added to `stats_history_` (two level std::map) as the value.

Reply:
`InternalStats->GetMapProperty` returns all cumulative counters in the format of `map<string, string>`. Of all the counters, only one is not in floating point form: `write_stall_micros_readable` and there is a related `write_stall_micros` so skipping it will not cause loss of information. Shall convert the map to `map<string, dobule>` to make it easier to calculate the delta?